filebeat/crawler/prospector.go

```diff
@@ -97,35 +100,89 @@ func (p *Prospector) Run(wg *sync.WaitGroup) {
 				return
 			case event := <-p.harvesterChan:
 				p.spoolerChan <- event
+				p.updateState(event.FileState)
 			}
 		}
 	}()
 
+	// Initial prospector run
+	p.prospectorer.Run()
+
 	for {
 		select {
 		case <-p.done:
 			logp.Info("Prospector stopped")
 			return
-		default:
+		case <-time.After(p.ProspectorConfig.ScanFrequencyDuration):
 			logp.Info("Run prospector")
 			p.prospectorer.Run()
 		}
 	}
 }
 
+func (p *Prospector) updateState(newState input.FileState) {
+
+	p.stateMutex.Lock()
+	defer p.stateMutex.Unlock()
```

Comment:
It is still needed as updatingState is called from a different go routine (receiving channels) then the running prospector itself. Would be nice to have a solution in the future where it is not needed anymore.